### ISSUE TITLE
Architecture of YOLOv11

### ISSUE THREAD

**Issue by ankanpy** [ultralytics/ultralytics#16559]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi @glenn-jocher,

Do you have any future plans to release a paper with the architecture improvement of yolov11 or any architectural diagram of the latest model?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @0xSynapse, thank you for your curiosity in the Ultralytics project 🚀!

For insights into our development process and future plans like YOLOv11, we suggest keeping an eye on our [Discussions](https://github.com/ultralytics/ultralytics/discussions) and community forums. While this isn't currently documented, you'll find updates shared there frequently.

If you're new to YOLO models or wish to refine your projects with existing architectures, visiting our [Docs](https://docs.ultralytics.com) is highly recommended. It includes detailed guides for various [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) cases.

Join the conversation in real-time on [Discord](https://ultralytics.com/discord) 🎧, or visit our [Subreddit](https://reddit.com/r/ultralytics) for community-driven discussions.

Remember, this is an automated response, and an Ultralytics engineer will reach out with more specific information soon. 😊

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This badge indicates if all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, which ensure the reliability and stability of our models across different platforms.

**Comment by Burhan-Q** [comment]
Hey there @0xSynapse I started working on a diagram for YOLO11 for the docs, but was not able to get it fully refined prior to launch. I'm planning to continue working on it later this week when I return home.

**Comment by glenn-jocher** [comment]
Thank you for your interest! Please keep an eye on our official channels for updates on new releases and architectures.

**Comment by glenn-jocher** [comment]
Thank you for your interest! Please stay tuned to our official channels for updates on YOLOv11's architecture.

**Comment by 0xSynapse** [comment]
It's too fast; great implementation!

Kudos to the Team!!

**Comment by glenn-jocher** [comment]
Thank you for the kind words! The credit goes to the entire YOLO community and the Ultralytics team. If you have any questions or need assistance, feel free to ask!

**Comment by AgungHari** [comment]
I have tried training YOLOv11 with a custom dataset. Is there an official architecture diagram available? I have tested it in the autonomous wheelchair project I am developing, and the results I obtained are remarkable, especially in terms of performance improvement. I hope to include these results in the paper I am currently working on.

**Comment by glenn-jocher** [comment]
I'm glad to hear about your success with YOLOv11! Currently, there's no official architecture diagram available, but please keep an eye on our updates for future releases.

**Comment by curtis18** [comment]
This is another unofficial architecture diagram from https://medium.com/@nikhil-rao-20/yolov11-explained-next-level-object-detection-with-enhanced-speed-and-accuracy-2dbe2d376f71

![image](https://github.com/user-attachments/assets/20320646-e1bd-4331-a3f9-37df7014ac8e)

**Comment by Burhan-Q** [comment]
@AbdullahOx unfortunately it doesn't look like we'll be releasing an official diagram.

**Comment by refdinal22** [comment]
@0xSynapse @AgungHari We made a video to explain the yolo11 architecture. Hope it helps. Link: https://youtu.be/L9Va7Y9UT8E?feature=shared

**Comment by Y-T-G** [comment]
@refdinal22 That looks nice. You can post it at https://community.ultralytics.com or https://reddit.com/r/Ultralytics

**Comment by AgungHari** [comment]
@refdinal22 Thank you very much for the video, it really helps to understand the YOLO11 architecture. I will watch it right away. Additionally, I would like to ask, what is inside the C3K2 block, and what is the fundamental difference compared to the C2F block in YOLOv8? Thanks to the ultralytics team for all the hard work!

**Comment by glenn-jocher** [comment]
Thank you for sharing the resource! It could be helpful for others exploring YOLO11.

**Comment by waqas-phc** [comment]
Hi @glenn-jocher Your last response was really helpful to me,

I was wondering if there will be any details on Architecture of Yolov11 Segmentation Model,
I know there are same as others with addition of Protonet Module . But I want to know from which layer the protonet layer is taken and from where I can access the mask coefficients of protonet masks.

any link to code will also be helpful.

thanks in advance.

**Comment by glenn-jocher** [comment]
@waqas-phc thank you for your question! The YOLO11 segmentation model uses the Protonet module, which is initialized in the `Segment` head class. The Protonet layer is typically applied to the feature map from the first input layer (`x[0]`) to generate mask prototypes, and mask coefficients can be accessed during the `forward` pass. You can find the relevant implementation in the [Segment class in `head.py`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/modules/head.py). Let us know if you need further assistance!

**Comment by waqas-phc** [comment]
Thanks man.

**Comment by waqas-phc** [comment]
Hi @glenn-jocher I know its a silly question but when we use `model.predict ` or `model(image.jpg)` it starts from [https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/model.py](url) Class. 

How the Model class is calling the Segment class from [https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/modules/head.py](head.py) .

**Comment by glenn-jocher** [comment]
The `Model` class in `model.py` initializes and manages the entire YOLO pipeline. When `model.predict` or `model(image.jpg)` is called, it uses the `Segment` class from `head.py` as part of the model architecture during the forward pass, where the head processes feature maps from the backbone. You can trace this connection in the model's initialization and forward methods. Let me know if you need further clarification!

**Comment by muhammadwaqas476074** [comment]
![image_2025-01-08_204254954](https://github.com/user-attachments/assets/83a7beb7-125c-46f4-8814-832c498abfea)
I looked at all the code files but could not find a single place where Segment is used for Initialization. 

Can you help in locating it

**Comment by muhammadwaqas476074** [comment]
Hi @glenn-jocher 
I am able to access all the predictions of yolo outputs from postprocessor but don't know what does these represent.

in Postprocessor it is clear that
`proto = preds[1][-1] if isinstance(preds[1], tuple) else preds[1]`
and
Mask will be computed by 
`masks = ops.process_mask_native(proto[i], pred[:, 6:], pred[:, :4], orig_img.shape[:2])`

But What About Other parameters of pred

Like 
we have 
`pred[0]
``pred[1]`


and their 
subparameters.


`Shape of preds[0]: torch.Size([1, 116, 6300])
Structure of preds[0]:
Shape: torch.Size([1, 116, 6300])
Number of elements: 730800
Data type: torch.float32
Device: cuda:0
Strides: (730800, 6300, 1)
Structure of preds[1]:
Length: 3
Element [0]:
  Length: 3
  Element [0]:
    Shape: torch.Size([1, 144, 60, 80])
    Number of elements: 691200
    Data type: torch.float32
    Device: cuda:0
    Strides: (691200, 4800, 80, 1)
  Element [1]:
    Shape: torch.Size([1, 144, 30, 40])
    Number of elements: 172800
    Data type: torch.float32
    Device: cuda:0
    Strides: (172800, 1200, 40, 1)
  Element [2]:
    Shape: torch.Size([1, 144, 15, 20])
    Number of elements: 43200
    Data type: torch.float32
    Device: cuda:0
    Strides: (43200, 300, 20, 1)
Element [1]:
  Shape: torch.Size([1, 32, 6300])
  Number of elements: 201600
  Data type: torch.float32
  Device: cuda:0
  Strides: (201600, 6300, 1)
Element [2]:
  Shape: torch.Size([1, 32, 120, 160])
  Number of elements: 614400
  Data type: torch.float32
  Device: cuda:0
  Strides: (614400, 19200, 160, 1)`

**Comment by muhammadwaqas476074** [comment]
I just wanna know what each element represent in `pred` and so that we can use them for different downstreaming tasks

**Comment by glenn-jocher** [comment]
In YOLOv11 segmentation models, `preds[0]` typically contains bounding box predictions, confidence scores, and mask coefficients, while `preds[1]` includes mask prototypes. For detailed downstream tasks, you can manipulate these elements after post-processing (e.g., non-max suppression). The segmentation-specific prototypes (`preds[1]`) can be processed with mask coefficients from `preds[0]` to generate final masks. You can refer to the [postprocess method in `segment/predict.py`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/segment/predict.py#L28) for implementation details. Let me know if you need further clarification!

**Comment by waqas-phc** [comment]
Hi @glenn-jocher Hope You are good. 
I am about to finish my work and I have some more questions.
**1. While in Object tracking is it possible to use the embeddings from P3, P4, P5 layers** 

> What I am doing is taking Mask from Protonet layer, scaling it and apply on P3, P4, P5 to extract all objects from all layers as ROI
Like this
`def extract_roi_with_mask(feature_map, scaled_mask, output_size=(4, 4)):
    """
    Extract the masked region of interest from the feature map.
    """
    if scaled_mask.numel() == 0 or feature_map.numel() == 0:
        return None

    # Apply the mask to the feature map
    masked_feature_map = feature_map * scaled_mask

    # Perform RoI Align (interpolation) to a fixed size
    roi_aligned = F.interpolate(masked_feature_map, size=output_size, mode='bilinear', align_corners=False)
    return roi_aligned`

. Then Applying Spatial Pyramid Pooling (SPP) Like this 
`def spp_pooling(roi_aligned, levels=[1, 2, 4]):
    """
    Apply Spatial Pyramid Pooling (SPP) to produce a fixed-size embedding.
    """
    if roi_aligned is None:
        return None  # Skip empty regions

    spp_features = []
    for level in levels:
        pool_size = (level, level)
        pooled = F.adaptive_avg_pool2d(roi_aligned, output_size=pool_size)
        spp_features.append(pooled.view(-1))  # Flatten each grid level

    return torch.cat(spp_features)`

Which will return us the embeddings of each object of Fix size. 

**2. Can the 32 masks of protonet layers  act as an Image imbedding of all classes (mean or sum of them) or the Protonet layer is just the shapes of detected objects with mask coefficients**
or 
**Can we use sum or mean of P3 as features embeddings**
> To be used in Robotics Vision down streaming task. where Agent has to interact with environment of different size objects

**Comment by glenn-jocher** [comment]
Hi @waqas-phc, thank you for your detailed query! 

1. Yes, it's possible to use embeddings from P3, P4, and P5 for downstream tasks like ROI extraction and SPP pooling. Your provided implementation is a valid approach for extracting and processing ROIs from feature maps using scaled masks. Ensure the scaling and alignment of the masks match the corresponding feature map resolutions for accurate extraction.

2. The 32 prototypes generated by the Protonet layer represent shared mask bases, which are combined with mask coefficients to reconstruct the instance masks. While these prototypes primarily encode spatial information about object shapes, they could potentially be aggregated (e.g., via mean or sum) to act as generalized embeddings. Similarly, features from P3, P4, or P5 can be aggregated to create embeddings, but their effectiveness depends on the task and domain-specific requirements. These features are rich in semantic and spatial information, making them suitable for tasks like robotic vision.

Feel free to explore and fine-tune these approaches for your downstream tasks, and let us know if you need further clarification!

**Comment by waqas-phc** [comment]
Thanks man for your help so far. Really means a lot to me.

**Comment by glenn-jocher** [comment]
You're very welcome! I'm glad I could help—feel free to reach out if you have more questions about YOLO or Ultralytics.

**Comment by Petros626** [comment]
@glenn-jocher what is the name of the YOLOv11 backbone?

**Comment by glenn-jocher** [comment]
YOLO11 uses Ultralytics’ own backbone (no separate external name); you can inspect it in the backbone section of the YOLO11 base YAML and use the Model YAML Configuration guide to interpret the modules. See the YOLO11 base YAML and the Model YAML Configuration guide. ([github.com](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/11/yolo11.yaml))

**Comment by DWSuryo** [comment]
Speaking of YOLOv11 architecture, do we have detailed view or image of each block's architecture like this one for YOLOv8? https://github.com/ultralytics/ultralytics/issues/189

**Comment by Petros626** [comment]
@DWSuryo I plan to create one based on the `.pptx` of this https://github.com/ultralytics/ultralytics/issues/189#issue-1527158137.

**Comment by Petros626** [comment]
> Thank you, everyone, for your input.
> 
> For those interested, I found this overview: https://arxiv.org/abs/2410.17725. I haven’t read it yet, but it might be useful.

for the architecture not.

**Comment by glenn-jocher** [comment]
We don’t currently maintain an official per-block architecture diagram for YOLO11 like the one in the [YOLOv8 architecture issue](https://github.com/ultralytics/ultralytics/issues/189); the most reliable “source of truth” is the [YOLO11 base YAML](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/11/yolo11.yaml) plus the [Model YAML Configuration guide](https://docs.ultralytics.com/guides/model-yaml-config/) and the module refs in [block modules (`block.py`)](https://docs.ultralytics.com/reference/nn/modules/block/) / [head modules (`head.py`)](https://docs.ultralytics.com/reference/nn/modules/head/)—and if you produce a `.pptx`/diagram, please share it in a new GitHub Discussion so it’s easy for the team/community to review (the arXiv link may be interesting context, but it’s not an official Ultralytics diagram).

**Comment by Petros626** [comment]
I tried my luck and created YOLO11 OBB, hopefully correct, but with the `.pptx` file, you can always suggest changes, beautification etc. I don't got the time to take care of changes, so it's up to the community.

<img width="3250" height="3402" alt="Image" src="https://github.com/user-attachments/assets/7df02eee-8fd3-4774-9499-64dc97a5d775" />

[YOLO11_OBB_architecture.pptx](https://github.com/user-attachments/files/24940366/YOLO11_OBB_architecture.pptx)

**Comment by glenn-jocher** [comment]
Thanks for sharing this—could you repost the `.pptx` (and note the exact checkpoint/config you built it from, plus the text output of `model.info()` for that model) in a new [Ultralytics GitHub Discussion](https://github.com/orgs/ultralytics/discussions) so it’s easier for others to review/iterate; for OBB-specific details it should align with the current [OBB task implementation](https://docs.ultralytics.com/tasks/obb/).

**Comment by Petros626** [comment]
I've build it from the `yolov11-obb.yaml` with basic specification of `nano`. https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/11/yolo11-obb.yaml

**Comment by glenn-jocher** [comment]
Got it—if the diagram was built from the `n` (nano) scale in [the `yolo11-obb.yaml` config](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/11/yolo11-obb.yaml), please also include the exact `ultralytics` version/commit you used and paste the text output of `model.info()` so others can verify the blocks match the current code.

**Comment by DWSuryo** [comment]
https://github.com/ultralytics/ultralytics/issues/16559#issuecomment-3811312588 by the way @Petros626 how about the YOLOv11 non-OBB? just omit the OBB block or are there other differences?

**Comment by Petros626** [comment]
> [#16559 (comment)](https://github.com/ultralytics/ultralytics/issues/16559#issuecomment-3811312588) by the way [@Petros626](https://github.com/Petros626) how about the YOLOv11 non-OBB? just omit the OBB block or are there other differences?

Non obb is the default detection model. Out of mind I can't say that for sure, but a look in the code and yaml will help you. When I remember correctly I changed just the head.

**Comment by glenn-jocher** [comment]
Yes—`yolo11-obb.yaml` is basically the standard YOLO11 backbone+neck with the detection head swapped for an `OBB` head that adds an angle branch (extra `angle` output and different decode/postprocess), so for “non-OBB” you typically just revert to the regular `Detect` head; the cleanest way to confirm is to diff the YAMLs in the [YOLO11 model config folder](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/cfg/models/11) and compare `Detect` vs `OBB` in the [head module reference](https://docs.ultralytics.com/reference/nn/modules/head/), plus paste `model.info()` for both to verify the last module differs.

**Comment by DWSuryo** [comment]
For upsampling, I searched for "upsample" words and I found:

* `nn/tasks.py` use `nn.Upsample()` for `attempt_load_weights()` function, which is for trainer engine.
* `nn/modules/block.py` use `nn.Upsample` for `Proto()` class.
* `utils/ops.py` has `if upsample` inside `process_,mask()` function, which also connects to Proto class, which apparently used for segmentation as I read.

Then in `ops.py`, I found `nn.functional.interpolate()` or `F.interpolate()`. So, I searched the function again and apparently, it's in `models/yolo/detect/train.py` inside `preprocess_batch()` inside `DetectionTrainer()` class. Correct me, is this where the upsampling happens for the YOLOv11 detection architecture?

Edit: oh sorry, I just found out in `yolov11.yaml` file especially Head section, it explicitly calls `nn.Upsample` with scale factor and nearest neighbor method. That's a bit straightforward than I thought.

**Comment by glenn-jocher** [comment]
You’re right: for YOLO11 *detection* the feature-map upsampling in the neck happens exactly where the model graph defines it—via the `nn.Upsample(scale_factor=2, mode="nearest")` entries in `yolo11.yaml`—while `DetectionTrainer.preprocess_batch()` is just resizing/formatting the *input images* for training (not the P3/P4/P5 FPN upsample); the `F.interpolate()` you found is primarily used in segmentation/mask paths (e.g., Proto/mask processing) rather than the core detect neck—if you want to double-check end-to-end, `model.info(detailed=True)` + the YAML interpreted via the [Model YAML Configuration guide](https://docs.ultralytics.com/guides/model-yaml-config/) is the most reliable way to trace where each upsample occurs.
